stdlib/runtime/numberUtils.gr

```diff
@@ -0,0 +1,1334 @@
+  }
+  WasmI32.toGrain(str): String
+}
+
+export let itoa64 = (value, radix) => {
+  if (WasmI32.ltS(radix, 2n) || WasmI32.gtS(radix, 36n)) {
+    fail "toString() radix argument must be between 2 and 36"
+  }
+
+  let mut value = value
+  let mut out = 0n
+
+  let sign = WasmI32.wrapI64(WasmI64.shrU(value, 63N))
+  if (sign != 0n) { value = WasmI64.sub(0N, value); }
+
+  if (WasmI64.eqz(value)) {
+    out = singleByteString(_CHAR_CODE_0);
+  } else if (radix == 10n) {
+    if (WasmI64.leU(value, _I32_MAX)) {
+      let val32    = WasmI32.wrapI64(value)
+      let decimals = decimalCount32(val32) + sign
+      out = allocateString(decimals)
+      utoa32_dec_core(out + 8n, val32, decimals);
+    } else {
+      let decimals = decimalCount64High(value) + sign
+      out = allocateString(decimals)
+      utoa64_dec_core(out + 8n, value, decimals);
+    }
+  } else if (radix == 16n) {
+    let decimals = WasmI32.shrU(63n - WasmI32.wrapI64(WasmI64.clz(value)), 2n) + 1n + sign
+    out = allocateString(decimals)
+    utoa64_hex_core(out + 8n, value, decimals);
+  } else {
+    let decimals = ulog_base(value, radix) + sign
+    out = allocateString(decimals)
+    utoa64_any_core(out + 8n, value, decimals, radix);
+  }
+  if (sign != 0n) WasmI32.store8(out, _CHAR_CODE_MINUS, 8n)
+  WasmI32.toGrain(out): String
+}
+
+let mut _K = 0n
+
+let umul64f = (u, v) => {
+  let u0 = WasmI64.and(u, 0xFFFFFFFFN)
+  let v0 = WasmI64.and(v, 0xFFFFFFFFN)
+
+  let u1 = WasmI64.shrU(u, 32N)
+  let v1 = WasmI64.shrU(v, 32N)
+
+  let l = WasmI64.mul(u0, v0)
+  let mut t = WasmI64.add(WasmI64.mul(u1, v0), WasmI64.shrU(l, 32N))
+  let mut w = WasmI64.add(WasmI64.mul(u0, v1), WasmI64.and(t, 0xFFFFFFFFN))
+
+  w = WasmI64.add(w, 0x7FFFFFFFN) // rounding
+
+  t = WasmI64.shrU(t, 32N)
+  w = WasmI64.shrU(w, 32N)
+
+  WasmI64.add(WasmI64.add(WasmI64.mul(u1, v1), t), w)
+}
+
+let umul64e = (e1, e2) => {
+  e1 + e2 + 64n // where 64 is significand size
+}
+
+let grisuRound = (buffer, len, delta, rest, ten_kappa, wp_w) => {
+  let mut lastp = buffer + len - 1n
+  let mut digit = WasmI32.load8U(lastp, 0n)
+  let mut rest = rest
+  while (
+    WasmI64.ltU(rest, wp_w) &&
+    WasmI64.geU(WasmI64.sub(delta, rest), ten_kappa) && (
+      WasmI64.ltU(WasmI64.add(rest, ten_kappa), wp_w) ||
+      WasmI64.gtU(WasmI64.sub(wp_w, rest), WasmI64.sub(WasmI64.add(rest, ten_kappa), wp_w))
+    )
+  ) {
+    digit -= 1n 
+    rest = WasmI64.add(rest, ten_kappa);
+  }
+  WasmI32.store8(lastp, digit, 0n)
+}
+
+let genDigits = (buffer, w_frc, w_exp, mp_frc, mp_exp, delta, sign) => {
+  let mut delta = delta
+  let one_exp = 0n - mp_exp
+  let one_frc = WasmI64.shl(1N, WasmI64.extendI32U(one_exp))
+  let mask    = WasmI64.sub(one_frc, 1N)
+
+  let mut wp_w_frc = WasmI64.sub(mp_frc, w_frc)
+
+  let mut p1 = WasmI32.wrapI64(WasmI64.shrU(mp_frc, WasmI64.extendI32U(one_exp)))
+  let mut p2 = WasmI64.and(mp_frc, mask)
+
+  let mut kappa = decimalCount32(p1)
+  let mut len = sign
+
+  let mut done = false
```

Comment:
Issue for early return? Then cleanup this stuff

Reply:
https://github.com/grain-lang/grain/issues/553